src/DataStructures/SliceIterator.cpp

```diff
@@ -37,11 +39,74 @@ void SliceIterator::reset() {
   slice_offset_ = 0;
 }
 
+template <size_t VolumeDim>
+std::pair<std::unique_ptr<std::pair<size_t, size_t>[], decltype(&free)>,
+          std::array<std::pair<gsl::span<std::pair<size_t, size_t>>,
+                               gsl::span<std::pair<size_t, size_t>>>,
+                     VolumeDim>>
+volume_and_slice_indices(const Index<VolumeDim>& extents) noexcept {
+  // Compute the slice indices once since this is (surprisingly) expensive
+  std::unique_ptr<std::pair<size_t, size_t>[], decltype(&free)> indices_buffer(
+      nullptr, &free);
+  // array over dim, pair over lower/upper, span<pair> over volume/boundary
+  std::array<std::pair<gsl::span<std::pair<size_t, size_t>>,
+                       gsl::span<std::pair<size_t, size_t>>>,
+             VolumeDim>
+      volume_and_slice_indices{};
+
+  size_t total_boundary_points = 0;
```

Comment:
As commented elsewhere, I think this should be named "half_number_boundary_points", and it can be made const if assigned by a lambda.